### ISSUE TITLE
Completely dynamic form creation

### ISSUE THREAD

**Issue by tashoecraft** [MrWolfZ/ngrx-forms#124]
I'm looking into ngrx-forms, however I may be blocked by the need to have completely dynamic form creation. During runtime we hit an api to get back json that dictates how the form should look and behave. Everything form input type, to validations, is controlled by the response. There are hundreds of thousands of these individual responses, so I can't possibly type out all of the possibilities.

Do you think this is something ngrx-forms could potentially support? I'd be willing to try and make it as I've run up against so many limits with the current reactive forms, if it's something that could potentially get merged in.

**Comment by MrWolfZ** [comment]
The dynamic input types, can they be objects and primitives or just primitives? From what you write the only problem implementing your scenario with ngrx-forms would be specifying the interface for the form value. If the input types can only be primitive you could just type all controls as string (or as union of all primitive types) and then evaluate the real type at runtime. This way you can get almost all features of ngrx-forms without any overhead. Alternatively you could type everything as `any` and perform type casts to the proper control types (e.g. group, array, control) based on your metadata, but this will make the code more complex.

If you could provide me with a sample response from your API I could show you a code example of how to create a form dynamically from that.

**Comment by tashoecraft** [comment]
So here's an example. It's a page of 5 questions, each question has a type property that dictates what type of question we should show. So this page has two different types, a standard input field, text_short, and a dropdown, "select_single". In total we have about 15+ question types. 

The select singles make requests to get lists of possible choices. So on initial load, country will get all possible countries, then when the user selects a country, state/province,county makes a request to get all results that would apply to the country, then city. 

The questions comes with validations, that we parse and apply to each question. Sometimes these validations need to look at other fields, which it looks like ngrx-forms does well (so awesome). Sometimes the entire page needs to be "repeated" so all questions get asked again in there own isolated group. 

Thanks for the help!

{
    "id":1,
    "validations":["required"],
      "questions":[
      {
        "id":2,
        "value":"Street Address (line 1)",
        "validations":["maxlength(100)","required"],
        "answer_type":"text_short",
      {
        "id":3,
        "value":"City",
        "validations":["maxlength(100)","required"],
        "answer_type":"text_short",
      },
      {
        "id":4,
        "value":"Zip",
        "validations":["maxlength(5)","required"],
        "answer_type":"text_short",
      },
      {
        "id": 5
        "value":"Country",
        "validations":["required"],
        "answers":[{"id":100}],
        "answer_type":"select_single",
        "list_identifier":"en_countries",
      },
      {
        "id":6,
        "value":"State/Province/County",
        "validations":["required"],
        "answers":[{"id":101}],
        "answer_type":"select_single",
        "list_identifier":"__NA__",
        "trigger_question":"5",
      },{
        "id":7,
        "value":"Closest Metropolitan Area",
        "validations":["required"],
        "answers":[{"id":102}],
        "answer_type":"select_single",
        "list_identifier":"__NA__",
        "trigger_question":"6",
        "user_answers":[]
      }
    ]
  }
}
![screen shot 2018-11-07 at 9 52 21 am](https://user-images.githubusercontent.com/10655908/48138696-dc00f700-e272-11e8-828c-9b82ca91e22b.png)

**Comment by MrWolfZ** [comment]
Sorry for the late response.

The first thing to observe in your use case is that the type of control (input, textarea, select etc.) does not matter for ngrx-forms. Only the data type of the value matters. In your case it seems to be fine to just use string everywhere. Therefore, your form value interface could look like this:

```typescript
export interface QuestionsFormValue {
  [questionId: string]: string;
}
```

So when you load the questions metadata from your API you can simply create an object implementing that interface and create the form state from that:

```typescript
const initialFormValue = questions.reduce((fv, q) => ({ ...fv, [q.id]: '' }), {} as QuestionsFormValue);
const initialFormState = createFormGroupState('some ID', initialFormValue);
```

Finally, for the validation there are multiple things you can do. One option would be to store the string definition of the validation on each question as a user-defined property and then inside of a validation function for the questions form state parse those definitions to proper functions. So the full code could look like this:

```typescript
import {
  AbstractControlState,
  createFormGroupState,
  setUserDefinedProperty,
  StateUpdateFns,
  updateGroup,
  updateRecursive,
  validate,
  ValidationErrors
} from 'ngrx-forms';

const questions = [
  {
    id: 1,
    validations: ['required'],
  },
];

export interface QuestionsFormValue {
  [questionId: string]: string;
}

const initialFormValue = questions.reduce((fv, q) => ({ ...fv, [q.id]: '' }), {} as QuestionsFormValue);
const initialFormState = createFormGroupState('some ID', initialFormValue);

const formStateValidationMetaDataUpdate = questions.reduce(
  (upd, q) => ({ ...upd, [q.id]: setUserDefinedProperty('validations', q.validations) }),
  {} as StateUpdateFns<QuestionsFormValue>
);

const formStateWithValidationMetadata = updateGroup(initialFormState, formStateValidationMetaDataUpdate);

const validateFormState = updateRecursive(s => s.userDefinedProperties.validations ? validateControl(s) : s);

function validateControl(control: AbstractControlState<any>) {
  const validationMetaData = control.userDefinedProperties.validations as string[];
  const validationFuncs = validationMetaData.map(parseValidationFunction);
  return validate(control, validationFuncs);
}

function parseValidationFunction(s: string): (value: any) => ValidationErrors {
  // TODO
  return () => ({});
}

// inside your reducer call the validation
const validatedFormState = validateFormState(formStateWithValidationMetadata);
```

**Comment by tashoecraft** [comment]
Not late at all and no need to apologize. Thanks for all of this! I really should have just seen that you had dynamic forms in your example app. Should be able to craft what I need.

**Comment by tashoecraft** [comment]
So I've gone through the docs pretty heavily and I'm still alittle confused about how to accomplish what I'm looking for.

To start I'm attempting to have a multi nested structure. The top layer will be a formArray, which will contain an array of formGroups, These formGroups will then contain the actual form field controls that will be linked up. This is because there are some pages of questions that can be repeated, so the entire form needs the ability to duplicate. I attempted to do something like this:

```
      let formGroup = createFormGroupState(innerQuestion.id, questions);
      let formArray = createFormArrayState(topQuestion.id, [formGroup]);
```

This appeared to work, until I looked down into the structure and all of the ngrx-forms values, touched, dirty, etc, were added to the controls of the bottom most formGroup as if they were individual form controls.

The docs are a little unclear about creating nested structures as nesting is only ever mentioned in the updateGroup section. 

While I haven't dug into it yet, the complexity is going to increase as every one of the formControls is going to have a custom value, which it looks like I'll be using the value boxing. The reasons for this are that I have to submit a whole bunch of additional information along with just the text value of an answer. 

I'm also having trouble figuring out what you wrote above fits into a standard reducer file. I'll need assemble the entire form inside reducer as an action will be dispatched with all of the form information during runtime.
